VariantToData:Fix float to integral conversion

sanmut opened the pull request:
When VariantToData is used to convert a variant of float to integer, it was not following G's general coercion rules. Fixing it to match the coercion rules and added some tests.

spathiwa commented on source/core/DualTypeConversion.cpp:
```diff
@@ -94,25 +94,24 @@ namespace Vireo
+        auto destEnumCount = static_cast<TDest>(destTypeRef->GetEnumItemCount());
+#pragma warning(push)
+#pragma warning(disable : 4018)  // Warning C4018 '>=': signed / unsigned mismatch
+        // It is safe to disable this warning because 'src' should be >= 0 here.
```
Please don't put compiler-specific pragmas in the code like this unless absolutely necessary, and if you do, but #if guards to test the platform before using them.
In this case, I think it would be better to change the code to `static_cast<TDest>(src >= destEnumCount)` than to disable the warning.

sanmut replied:
Since clang seems to ignore the pragmas, I left them in. I agree it is a good idea not to do compiler specific pragmas without guards. OTOH, casting either src or destEnumCount with other's type won't work as the C++ cast will do the normal roll-over and not stop at the target type's max value. OTOH, enum supports only UInt8,16,32, so it is safe to cast both to UInt32 to compare -- which would avoid the compiler warning. I updated the PR with this.

spathiwa commented on source/core/DualTypeConversion.cpp:
```diff
@@ -94,25 +94,24 @@ namespace Vireo
     }
 
-    template<typename T> void ApplyNumeric(T valueX, TypeRef typeRefY, void* pDataY)
+    template<typename T> void ApplyIntegralToNumeric(T valueX, TypeRef typeRefY, void* pDataY)
```
'Integral' reads weird to me.  Why not just 'Integer'?

sanmut replied:
I went with "Integral" to imply Integer and enum. Enum is sort of an integral type with underlying integer. If you feel strongly about it, I can fix this to make it 'Integer'.

spathiwa replied:
I don't feel that strongly, if you want to leave it, it's fine.

sanmut replied:
I added some comments about the convention used in the file for Integer/Integral/Numeric.

sanmut commented:
While adding more test cases to make sure the conversion works correctly, I also found IsA does not do the right thing in comparing enum type against an integer type. So, I fixed that and added tests for the same too.

sanmut commented:
Addressed Sid's feedback on renaming ConvertFromEnum method and leaving comment.
